Test runner blows up on first run: [ERROR] TestRunnerChildProcessAdapter

<!---
Thanks for filing an issue 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have asked the same thing before!

If you are having trouble running stryker, please fill in the template below.
If you have a feature request, feel free to ignore the template.
-->

### Summary

I'm trying to use Stryker for the first time on https://gitlab.com/NodeGuy/channel and it appears to break when it hits a mutant (I'm guessing):

```shell
$ stryker run
[2018-07-18 18:00:11.550] [INFO] ConfigReader - Using stryker.conf.js in the current working directory.
[2018-07-18 18:00:11.952] [INFO] InputFileResolver - Found 1 of 20 file(s) to be mutated.
[2018-07-18 18:00:11.955] [INFO] InitialTestExecutor - Starting initial test run. This may take a while.
[2018-07-18 18:00:14.545] [INFO] InitialTestExecutor - Initial test run succeeded. Ran 49 tests in 2 seconds (net 100 ms, overhead 85 ms).
[2018-07-18 18:00:14.962] [INFO] Stryker - 252 Mutant(s) generated
[2018-07-18 18:00:15.055] [INFO] SandboxPool - Creating 8 test runners (based on CPU count)
Mutation testing  [======                                            ] 12% (ETC 30.1s) 31/252 tested (4 survived)[2018-07-18 18:00:21.788] [ERROR] TestRunnerChildProcessAdapter - /Users/david/Documents/projects/channel/.stryker-tmp/sandbox7828783/lib/index.js:85
      const { cancelled, resolve } = pushes[resolvedIndex];
              ^

TypeError: Cannot destructure property `cancelled` of 'undefined' or 'null'.
    at Immediate.processOrders (/Users/david/Documents/projects/channel/.stryker-tmp/sandbox7828783/lib/index.js:85:44)
    at runCallback (timers.js:763:18)
    at tryOnImmediate (timers.js:734:5)
    at processImmediate (timers.js:716:5)

[2018-07-18 18:00:21.797] [ERROR] TestRunnerChildProcessAdapter - Child process exited with non-zero exit code 1. Last 10 message from the child process were: 
	[2018-07-18 18:00:15.934] [DEBUG] PluginLoader - Loading stryker-* from /Users/david/Documents/projects/channel/node_modules
[2018-07-18 18:00:15.942] [DEBUG] PluginLoader - Loading plugins stryker-javascript-mutator (matched with expression stryker-*)
[2018-07-18 18:00:15.943] [DEBUG] PluginLoader - Loading plugins stryker-mocha-framework (matched with expression stryker-*)
[2018-07-18 18:00:15.943] [DEBUG] PluginLoader - Loading plugins stryker-mocha-runner (matched with expression stryker-*)

	[2018-07-18 18:00:15.943] [DEBUG] PluginLoader - Loading plugins stryker-javascript-mutator

	[2018-07-18 18:00:17.309] [DEBUG] PluginLoader - Loading plugins stryker-mocha-framework

	[2018-07-18 18:00:17.326] [DEBUG] PluginLoader - Loading plugins stryker-mocha-runner

	[2018-07-18 18:00:17.590] [DEBUG] IsolatedTestRunnerAdapterWorker - Changing current working directory for this process to /Users/david/Documents/projects/channel/.stryker-tmp/sandbox7828783

[2018-07-18 18:00:21.799] [ERROR] StrykerCli - an error occurred Test runner child process exited with non-zero exit code 1
```

### Stryker config

<!--- Please place your stryker config below. Feel free to change paths in the files and mutate arrays if you cannot share them. -->

```js
module.exports = function(config) {
  config.set({
    testRunner: "mocha",
    mutator: "javascript",
    transpilers: [],
    reporter: ["clear-text", "progress"],
    testFramework: "mocha",
    coverageAnalysis: "perTest",
    mutate: ["lib/**/*.js"]
  });
};

```

### Stryker environment

<!-- Please list your stryker plugins + versions here (output of `npm ls | grep stryker`) -->

```
$ npm ls | grep stryker
├─┬ stryker@0.24.2
├─┬ stryker-api@0.17.3
├─┬ stryker-javascript-mutator@0.7.3
├─┬ stryker-mocha-framework@0.10.3
└─┬ stryker-mocha-runner@0.12.3
```

### Your Environment
| software         | version(s)
| ---------------- | -------  
| node             |  v9.11.1
| npm              | 5.6.0
| Operating System | macOS 10.13.4

--- a/packages/stryker/src/isolated-runner/RetryDecorator.ts
+++ b/packages/stryker/src/isolated-runner/RetryDecorator.ts
@@ -33,24 +33,18 @@ export default class RetryDecorator extends TestRunnerDecorator {
     return isErrnoException(error) && error.code === BROKEN_PIPE_ERROR_CODE;
   }
 
-  private tryRun(options: RunOptions, retriesLeft = 2, lastError?: any) {
-    if (retriesLeft > 0) {
-      this.innerRunner.run(options).then(result =>
-        this.currentRunTask.resolve(result),
-        rejectReason => {
-          if (this.innerProcessIsCrashed(rejectReason)) {
-            this.recover().then(
-              () => this.tryRun(options, retriesLeft - 1, rejectReason),
-              reason => this.currentRunTask.reject(reason));
-          } else {
-            // Oops... not intended to catch this one
-            this.currentRunTask.reject(rejectReason);
-          }
-        });
+  private async tryRun(options: RunOptions, attemptsLeft = 2, lastError?: any) {
+    if (attemptsLeft > 0) {
+      try {
+        let result = await this.innerRunner.run(options);
+        this.currentRunTask.resolve(result);
+      } catch (error) {
+        await this.recover();
+        await this.tryRun(options, attemptsLeft - 1, error);
+      }
     } else {
-      this.recover().then(
-        () => this.currentRunTask.resolve({ status: RunStatus.Error, errorMessages: [ERROR_MESSAGE + errorToString(lastError)], tests: [] }),
-        (reason) => this.currentRunTask.reject(reason));
+      await this.recover();
+      this.currentRunTask.resolve({ status: RunStatus.Error, errorMessages: [ERROR_MESSAGE + errorToString(lastError)], tests: [] });
     }
   }
 

--- a/packages/stryker/test/integration/isolated-runner/ResilientTestRunnerFactory.it.ts
+++ b/packages/stryker/test/integration/isolated-runner/ResilientTestRunnerFactory.it.ts
@@ -63,25 +63,25 @@ describe('ResilientTestRunnerFactory integration', function () {
   }
 
   it('should be able to receive a regex', async () => {
-    sut = arrangeSut('discover-regex');
+    arrangeSut('discover-regex');
     const result = await actRun();
     expect(result.status).eq(RunStatus.Complete);
   });
 
   it('should pass along the coverage result from the test runner behind', async () => {
-    sut = arrangeSut('coverage-reporting');
+    arrangeSut('coverage-reporting');
     const result = await actRun();
     expect(result.coverage).eq('realCoverage');
   });
 
   it('should pass along the run result', async () => {
-    sut = arrangeSut('direct-resolved');
+    arrangeSut('direct-resolved');
     const result = await actRun();
     expect(result.status).eq(RunStatus.Complete);
   });
 
   it('should try to report coverage from the global scope, even when the test runner behind does not', async () => {
-    sut = arrangeSut('direct-resolved');
+    arrangeSut('direct-resolved');
     const result = await actRun();
     expect(result.coverage).eq('coverageObject');
   });
@@ -118,7 +118,7 @@ describe('ResilientTestRunnerFactory integration', function () {
   });
 
   it('should be able to run twice in quick succession', async () => {
-    sut = arrangeSut('direct-resolved');
+    arrangeSut('direct-resolved');
     await actRun();
     const result = await actRun();
     expect(RunStatus[result.status]).eq(RunStatus[RunStatus.Complete]);
@@ -130,14 +130,14 @@ describe('ResilientTestRunnerFactory integration', function () {
   });
 
   it('should change the current working directory to the sandbox directory', async () => {
-    sut = arrangeSut('verify-working-folder');
+    arrangeSut('verify-working-folder');
     const result = await actRun();
     expect(result.errorMessages).undefined;
   });
 
   it('should be able to recover from any crash', async () => {
     // time-bomb will crash after 100 ms
-    sut = arrangeSut('time-bomb');
+    arrangeSut('time-bomb');
     await sleep(101);
     const result = await actRun();
     expect(RunStatus[result.status]).eq(RunStatus[RunStatus.Complete]);

--- a/packages/stryker/test/unit/isolated-runner/RetryDecoratorSpec.ts
+++ b/packages/stryker/test/unit/isolated-runner/RetryDecoratorSpec.ts
@@ -60,26 +60,22 @@ describe('RetryDecorator', () => {
       return expect(result).to.eventually.eq(expectedResult);
     });
 
-    it('should pass through non-crash related rejects', () => {
+    it('should retry on a new test runner if a reject appears', () => {
       testRunner1.run.rejects(new Error('Error'));
-      return expect(sut.run(options)).to.be.rejectedWith('Error');
-    });
-
-    it('should retry on a new test runner if a crash related reject appears', () => {
-      testRunner1.run.rejects(brokenPipeError);
       testRunner2.run.resolves(expectedResult);
       return expect(sut.run(options)).to.eventually.eq(expectedResult);
     });
 
-    it('should retry at most 3 times before rejecting', () => {
-      testRunner1.run.rejects(brokenPipeError);
-      testRunner2.run.rejects(brokenPipeError);
-      testRunner3.run.rejects(brokenPipeError);
-      testRunner4.run.rejects(brokenPipeError);
+    it('should retry at most 1 times before rejecting', () => {
+      const finalError = new Error('Error');
+
+      testRunner1.run.rejects(new Error('Error'));
+      testRunner2.run.rejects(finalError);
+
       return sut.run(options).then((runResult: RunResult) => {
         expect(runResult.status).to.be.eq(RunStatus.Error);
         expect(runResult.errorMessages).to.be.deep.eq(['Test runner crashed. Tried twice to restart it without any luck. Last time the error message was: '
-          + errorToString(brokenPipeError)]);
+        + errorToString(finalError)]);
         expect(availableTestRunners).to.have.lengthOf(0);
       });
     });

